### ISSUE TITLE
Importing the astropy.table module prevents from setting a multiprocessing start method (caused by numcodecs in io.fits)

### ISSUE THREAD

**Issue by mboquien** [astropy/astropy#16351]
### Description

Importing the `astropy.table` triggers an exception if one uses `multiprocessing.set_start_method()` afterwards

### Expected behavior

Importing the `astropy.table` does not trigger an exception if one uses `multiprocessing.set_start_method()` afterwards

### How to Reproduce

The following code reproduces the issue.

```python
import multiprocessing as mp
import astropy.table

mp.set_start_method('spawn')
```

It generates the following backtrace
```
Traceback (most recent call last):
  File "/home/mederic/scratch/start_method/test.py", line 4, in <module>
    mp.set_start_method('spawn')
  File "/usr/lib/python3.12/multiprocessing/context.py", line 247, in set_start_method
    raise RuntimeError('context has already been set')
RuntimeError: context has already been set
```

### Versions

```
Linux-6.8.7-arch1-1-x86_64-with-glibc2.39
Python 3.12.3 (main, Apr 23 2024, 09:16:07) [GCC 13.2.1 20240417]
astropy 6.0.0
Numpy 1.26.4
pyerfa 2.0.1.1
Scipy 1.13.0
Matplotlib 3.8.3
```

**Comment by mhvk** [comment]
That's weird - I cannot reproduce this (on Debian, astropy 6.0.0, python 3.11.8; rest shouldn't matter). Could it be something with python 3.12?

**Comment by mboquien** [comment]
In my case switching python 3.12 definitely triggers the bug. However I have been receiving sporadic reports of this issue with my code. While I was not always provided with the full information, one of the reporters had it on Python 3.11.8 on the win-amd64 platform.

**Comment by neutrinoceros** [comment]
I couldn't reproduce either on macOS with Python 3.12.2 + astropy 7.0dev

**Comment by astrofrog** [comment]
@mboquien just out of curiosity do you have dask installed? What is the output of 'pip freeze'?

**Comment by mboquien** [comment]
Dask was installed but removing it does not solve the issue. To follow the idea, I have also removed many of the optional astropy packages, to no avail. I attach the [list of packages](https://github.com/astropy/astropy/files/15141880/packages.txt) if it can help.

**Comment by mboquien** [comment]
I went down the rabbit hole a little more and I believe I have found the culprit in the form of the [numcodecs package](https://pypi.org/project/numcodecs/).

Now I am unsure whether the issue is with numcodecs itself or in the way astropy uses it.

**Comment by mhvk** [comment]
> Now I am unsure whether the issue is with numcodecs itself or in the way astropy uses it.

I don't think astropy uses it, not directly at least.
...
Just tries installing astropy in a clean environment, and I don't see `numcodecs` among the packages, not even when I do `pip install astropy[all]`.

**Comment by mboquien** [comment]
I looked at that with `git grep` and [Astropy uses it optionally](https://github.com/astropy/astropy/blob/main/astropy/io/fits/hdu/compressed/_codecs.py#L19).

**Comment by mhvk** [comment]
Ah, I'm silly, I did my `git grep` in an older branch I happened to be on...  But at least we now know the culprit and can understand why @neutrinoceros and I could not reproduce it. One lesson perhaps is that we should not try and use packages that are not also listed in our dependencies. And if things really work without them, perhaps it is better just not to use them at all?

Anyway, back on topic: it would now seem that the bug is with `numcodecs` rather than with astropy, since in a clean environment with just `numcodecs` installed, I can reproduce your error:
```
>>> import multiprocessing as mp
>>> import numcodecs
>>> mp.set_start_method('spawn')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.11/multiprocessing/context.py", line 247, in set_start_method
    raise RuntimeError('context has already been set')
RuntimeError: context has already been set
```

**Comment by mboquien** [comment]
Indeed, importing numcodecs directly triggers the issue, which [I have reported to them](https://github.com/zarr-developers/numcodecs/issues/522). Hopefully it will be addressed soon. In the meantime I know what to do to work around the issue. This being said, as a former Astropy packager for Archlinux, I think numcodecs should officially be listed as an optional dependency. Hidden dependencies can make issues more difficult to debug. In any case, thank you very much for your help!

**Comment by neutrinoceros** [comment]
Thanks for giving us proper motivation to go hunting for rogue optional dependencies in astropy, I've been meaning to do this systematically but lacked a concret reason to !

**Comment by astrofrog** [comment]
Maybe worth trying out https://github.com/tweag/FawltyDeps !

**Comment by mhvk** [comment]
@astrofrog - the particular use of `numcodecs` probably needs a bit of thought. From the comment, this was added to help people who use codecs elsewhere, which I think is a very good idea. But I think we do need to document and test it - and perhaps make it an explicit opt-in so we do not have to add it to astropy's dependencies - I'm not sure that we want dependencies where nothing we do depends on them, but rather we add functionality; that seems the wrong way around.

**Comment by pllim** [comment]
Re: 

https://github.com/astropy/astropy/blob/0c06ac3e145a2b83480be4a596c271f4795ee560/astropy/io/fits/hdu/compressed/_codecs.py#L19-L29

Is it worth adding an option for user to disable numcodecs? 🤷‍♀️

**Comment by saimn** [comment]
numcodecs creates a Lock at import time (which is not a great idea maybe and could probably be avoided) which triggers the creation of the default context in multiprocessing. Then changing the start method is not possible:
```
In [1]: import multiprocessing

In [2]: multiprocessing.Lock()
Out[2]: <Lock(owner=None)>

In [3]: multiprocessing.set_start_method("spawn")
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[3], line 1
----> 1 multiprocessing.set_start_method("spawn")

File ~/.pyenv/versions/3.12.2/lib/python3.12/multiprocessing/context.py:247, in DefaultContext.set_start_method(self, method, force)
    245 def set_start_method(self, method, force=False):
    246     if self._actual_context is not None and not force:
--> 247         raise RuntimeError('context has already been set')
    248     if method is None and force:
    249         self._actual_context = None

RuntimeError: context has already been set
```

But this is also due to the way multiprocessing works, `set_start_method` should be called at the very beginning otherwise this issue could happen with other libraries. And there is an alternative which is to create a new context with `get_context("spawn")`:
https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods

> [set_start_method()](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method) should not be used more than once in the program.
Alternatively, you can use [get_context()](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.get_context) to obtain a context object. Context objects have the same API as the multiprocessing module, and allow one to use multiple start methods in the same program.

So I would not classify this as a bug, though it would be better if numcodecs delay the creation of the lock.

**Comment by mboquien** [comment]
Thank you very much for the explanation. However, from my naive end-user point of view, I would much prefer not to have to use a workaround to use `astropy.table` with `multiprocessing`. I understand it should ideally be fixed in `numcodecs` directly. So far there has been [no activity on the `numcodecs` bug report](https://github.com/zarr-developers/numcodecs/issues/522) so it may take a while longer before the issue is fixed on their side.

**Comment by neutrinoceros** [comment]
> numcodecs creates a Lock at import time (which is not a great idea maybe and could probably be avoided)

I just opened a PR to address this upstream https://github.com/zarr-developers/numcodecs/pull/615

**Comment by mhvk** [comment]
I think I agree with @mboquien here (and hence removed the `close?` label). Our "helping" for a scenario that is unclear whether it is actually used (and certainly is not tested) made for a really hard to diagnose bug. I think we have a few options:
1. Just remove the `numcodecs` stuff altogether (or make it version dependent on having @neutrinoceros's patch)
2. Follow @pllim's suggestion of making it a configurable option to use the codecs. Its default could be version dependent.

For (2) or any other solution that keeps our current use, we should actually have tests and `numcodecs` should be listed as an optional dependency (though perhaps that can be for 7.0.1...)

**Comment by neutrinoceros** [comment]
The patch was just merged upstream. I'm also happy to contribute the testing + configurability as suggested here.

**Comment by neutrinoceros** [comment]
This is solved by the release of numcodecs 0.14.0
xref https://numcodecs.readthedocs.io/en/stable/release.html#release-0-14-0
